keps/sig-node/0029-2018-09-19-node-fencing.md

```diff
@@ -0,0 +1,223 @@
+   avoided.
+
+1. Implementations should indicate a Node needs to be fenced with the `NeedsFencing` NodeCondition.
+
+   Admins may also set NodeCondition `FencingRequired` to `True` to manually
+   trigger fencing.
+
+   Once fencing has been initiated, the NodeCondition `FencingComplete` should
+   be set to `False`.
+
+1. Implementations should indicate a Node has been sucessfully fenced by
+   setting the NodeCondition `FencingComplete` to `True`.
+
+   Fencing implementations must only set `FencingComplete`  once it confirms
+   any workloads on the node have been isolated from the network and any
+   shared resources.  Most commonly this will mean the node has been powered
+   off.
+
+1. Implementations should return Nodes to a "normal" state by removing the
+   `FencingComplete` and `FencingRequired` NodeConditions.
+
+   This could occur periodically or more likely, when the NodeConditions that
+   were determined to require fencing are no longer present.
+
+1. Kubernetes Controllers should look for `FencingRequired` = `True` **and**
+   `FencingComplete` = `True` to determine that it is safe to forcefully recover
+   resources.
+
+### Implementation Details/Notes/Constraints
+
+#### Node Conditions
+
+As setting and unsetting the `NodeConditions` form the primary contract between
+implementations and the rest of the system, implementations should consider
+failure modes that include or co-incide with interruptions to the service
+network between nodes and `apiservers`.
+
+For additional inputs to the repair logic, we rely on NPD which runs locally
+on each node. In the presence of a service network failure, NPD would not be
+able to set additional `NodeConditions`. However the fact that the node is
+disconnected should already be sufficient to trigger fencing of the affected
+nodes.
+
+Since the fencing implementation is responsible for setting and clearing
+`FencingTriaged`, `FencingRequired`, `FencingComplete`,  any disruption to the
+service network between it and the `apiservers` is also important to consider.
+
+In our implementation there are multiple copies that vote for a leader, the
+presence of a service network failure affecting the elected controller results
+in it being unable to renew its lease and one of its peers (on another node
+that by definitiion still has connectivity) taking over and being able to set
+`NodeConditions`.
+
+In the presence of a total failure, then we shouldn't be fencing anyway so
+there would be no need to set any `NodeConditions`.
+
+#### Power Management
+
+Since the most common form of fencing involves powering off nodes, co-
+ordination with other entities that may be involved with power management
+should be considered...
+
+TBC
+
+### Risks and Mitigations
+
+TBA
+
+## Graduation Criteria
+
+TBA
+
+## Implementation History
+
+- Initial version: 2018-09-19
+
+## Drawbacks
+
+* In deleting the Node object, we are also removing the history of events that
```

Comment:
Most of this doc does not talk about delete ... ?

Reply:
Right, this paragraph is missing some much needed context.

The rest of the document covers how implementations should identify nodes in need of fencing and indicate progress towards the node being made safe.

However once the node is safe, something must use that information to convey to the rest of the system that affected workloads are now safe to run elsewhere.

One approach would be having the fencing logic delete Node objects.
The other extreme is for every controller to monitor for the new Node Conditions and react appropriately.

I do not yet have a good sense of which part of the spectrum would be preferred.